pkg/controller/database/database_schema_controller.go

```diff
@@ -0,0 +1,207 @@
+		return reconcile.Result{}, errors.Wrap(err, "failed to open connection to mysql")
+	}
+	defer db.Close()
+
+	db.SetMaxIdleConns(0)
+	db.SetMaxOpenConns(1)
+
+	query := `SELECT default_character_set_name, default_collation_name FROM information_schema.SCHEMATA 
+	WHERE schema_name = ?`
+	row := db.QueryRow(query, cfg.DBName)
+
+	var defaultCharset, defaultCollation string
+	if err := row.Scan(&defaultCharset, &defaultCollation); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to scan")
+	}
+
+	if databaseInstance.Spec.Connection.Mysql.DefaultCharset == defaultCharset && databaseInstance.Spec.Connection.Mysql.Collation == defaultCollation {
+		logger.Debug("no action needed as the charset and collation already match")
+		return reconcile.Result{}, nil
+	}
+
+	// if they are different, it might be the server default, but that's hard to tell from a connection
+	// if the user passed "", then we want the database set to the default
+	// https://dev.mysql.com/doc/refman/8.0/en/charset-database.html  <-- see this for application
+	query = `show variables like 'collation_server'`
+	row = db.QueryRow(query)
+
+	var variableName string
+	var defaultServerCollation, defaultServerCharset string
+	if err := row.Scan(&variableName, &defaultServerCollation); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to scan server default collation")
+	}
+
+	query = `show variables like 'character_set_server'`
+	row = db.QueryRow(query)
+	if err := row.Scan(&variableName, &defaultServerCharset); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to scan server default charset")
+	}
+
+	/*
+		from https://dev.mysql.com/doc/refman/8.0/en/charset-database.html
+
+		MySQL chooses the database character set and database collation in the following manner:
+
+			If both CHARACTER SET charset_name and COLLATE collation_name are specified, character set charset_name and collation collation_name are used.
+
+			If CHARACTER SET charset_name is specified without COLLATE, character set charset_name and its default collation are used. To see the default collation for each character set, use the SHOW CHARACTER SET statement or query the INFORMATION_SCHEMA CHARACTER_SETS table.
+
+			If COLLATE collation_name is specified without CHARACTER SET, the character set associated with collation_name and collation collation_name are used.
+
+			Otherwise (neither CHARACTER SET nor COLLATE is specified), the server character set and server collation are used.
+
+	*/
+
+	if databaseInstance.Spec.Connection.Mysql.DefaultCharset != "" && databaseInstance.Spec.Connection.Mysql.Collation != "" {
+		// If both CHARACTER SET charset_name and COLLATE collation_name are specified, character set charset_name and collation collation_name are used.
+		query = fmt.Sprintf("alter database `%s` character set '%s' collate '%s'", cfg.DBName, databaseInstance.Spec.Connection.Mysql.DefaultCharset, databaseInstance.Spec.Connection.Mysql.Collation)
```

Comment:
any reason you use fmt rather than prepared statements here? does alter not support prepared statements?

Reply:
alter table is supported, but alter database is not https://dev.mysql.com/doc/refman/8.0/en/sql-prepared-statements.html